bindings/py/cpp_src/bindings/encoders/py_SimHashDocumentEncoder.cpp

```diff
@@ -0,0 +1,189 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU Affero Public License along with
+ * this program.  If not, see http://www.gnu.org/licenses.
+ * -------------------------------------------------------------------------- */
+
+/** @file
+ * py_SimHashDocumentEncoder.cpp
+ * @since 0.2.3
+ */
+
+#include <bindings/suppress_register.hpp>  //include before pybind11.h
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include <htm/encoders/SimHashDocumentEncoder.hpp>
+
+namespace py = pybind11;
+
+using namespace htm;
+
+
+namespace htm_ext {
+
+  using namespace htm;
+
+  void init_SimHashDocumentEncoder(py::module& m)
+  {
+    /**
+     * Parameters
+     */
+    py::class_<SimHashDocumentEncoderParameters>
+      py_SimHashDocumentEncoderParameters(m, "SimHashDocumentEncoderParameters",
+R"(
+Parameters for the SimHashDocumentEncoder.
+)");
+
+    py_SimHashDocumentEncoderParameters.def(py::init<>());
+
+    py_SimHashDocumentEncoderParameters.def_readwrite("activeBits",
+      &SimHashDocumentEncoderParameters::activeBits,
+R"(
+This is the number of true bits in the encoded output SDR. The output encoding
+will have a distribution of this many 1's. Specify only one of: activeBits
+or sparsity.
+)");
+
+    py_SimHashDocumentEncoderParameters.def_readwrite("size",
+      &SimHashDocumentEncoderParameters::size,
+R"(
+This is the total number of bits in the encoded output SDR.
+)");
+
+    py_SimHashDocumentEncoderParameters.def_readwrite("sparsity",
+      &SimHashDocumentEncoderParameters::sparsity,
+R"(
+This is an alternate way (percentage) to specify the the number of active bits.
+Specify only one of: activeBits or sparsity.
+)");
+
+    py_SimHashDocumentEncoderParameters.def_readwrite("tokenSimilarity",
+      &SimHashDocumentEncoderParameters::tokenSimilarity,
+R"(
+This allows similar tokens ("cat", "cats") to also be represented similarly,
+at the cost of document similarity accuracy. Default is FALSE (providing better
+document-level similarity, at the expense of token-level similarity).
+
+Results are heavily dependent on the content of your input data.
+
+If TRUE: Similar tokens ("cat", "cats") will have similar influence on the
+  output simhash. This benefit comes with the cost of a reduction in
+  document-level similarity accuracy.
+
+If FALSE: Similar tokens ("cat", "cats") will have individually unique and
+  unrelated influence on the output simhash encoding, thus losing token-level
+  similarity and increasing document-level similarity.
+)");
+
+
+    /**
+     * Class
+     */
+    py::class_<SimHashDocumentEncoder> py_SimHashDocumentEncoder(m,
+      "SimHashDocumentEncoder",
+R"(
+Encodes a document text into a distributed spray of 1's.
+
+The SimHashDocumentEncoder encodes a document (array of strings) value into an
```

Comment:
Yup!  Documents can be any # of tokens greater than 0.

I've improved the docs about this here: https://github.com/htm-community/htm.core/pull/603/commits/e40ee2f6fe166e1093cd3ea320aa3b530d9de591

Quick SimHash description:
* Each token gets a binary hash digest (array values: 0 or 1)
    * The 0's in those binsets get changed to -1  (array values: -1 or +1)
* Sum all the binary columns from all tokens
    * Take the indices of the top sums as our sparse ON bits.
    * Since we're always just picking off the top, the actual sum value does not matter.
        * So, the sum value of each bit column can be very high (lots of tokens in document) or low (just a few tokens per document). Works the same either way.

Good SimHash video:
https://www.youtube.com/watch?v=gnraT4N43qo